bindings/go/blob/buffered_reader.go

```diff
@@ -0,0 +1,137 @@
+package blob
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"github.com/opencontainers/go-digest"
+)
+
+// NewEagerBufferedReader creates a new EagerBufferedReader instance.
+func NewEagerBufferedReader(r io.Reader) *EagerBufferedReader {
+	return &EagerBufferedReader{
+		reader:    r,
+		mediaType: "application/octet-stream",
+	}
+}
+
+// EagerBufferedReader is a reader that can calculate its digest and size
+// By eagerly loading the data, and proxying a buffer for it in memory.
+// This can lead to extraordinary memory usage for large files, so one should
+// be careful when using this without considering the potential size of the data.
+type EagerBufferedReader struct {
+	mu        sync.RWMutex
+	buf       *bytes.Buffer
+	digest    string
+	size      uint64
+	loaded    bool
+	reader    io.Reader
+	mediaType string
+}
+
+var (
+	// By buffering the data, we can calculate the size and digest of the data
+	// from the buffer!
+	_ SizeAware   = &EagerBufferedReader{}
+	_ DigestAware = &EagerBufferedReader{}
+
+	// We can also set the media type of the buffer if needed
+	_ MediaTypeAware = &EagerBufferedReader{}
+)
+
+func (b *EagerBufferedReader) LoadEagerly() error {
+	if b.Loaded() {
+		return nil
+	}
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.buf == nil {
+		b.buf = &bytes.Buffer{}
+	}
+	if b.reader != nil {
+		b.reader = io.MultiReader(b.buf, b.reader)
+	}
+	dig, err := digest.FromReader(io.TeeReader(b.reader, b.buf))
+	if err != nil {
+		return err
+	}
+	b.digest = dig.String()
+	if newSize := uint64(b.buf.Len()); newSize > b.size { //nolint:gosec // G115 can never be negative
+		b.size = newSize
+	}
+	b.loaded = true
+	return nil
+}
+
+func (b *EagerBufferedReader) Loaded() bool {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
```

Comment:
Why is the lock necessary? We neither load nor alter the data

Reply:
I guess to make sure that the object is not processed by `b.LoadEagerly()` while calling `Loaded()`